src/analysis/seminorm.lean

```diff
@@ -255,26 +256,50 @@ variables [add_monoid E]
 section has_scalar
 variables [has_scalar 𝕜 E]
 
-instance : inhabited (seminorm 𝕜 E) :=
-⟨{ to_fun    := λ _, 0,
-   smul'     := λ _ _, (mul_zero _).symm,
-   triangle' := λ x y, by rw add_zero }⟩
+instance fun_like : fun_like (seminorm 𝕜 E) E (λ _, ℝ) :=
+{ coe := seminorm.to_fun, coe_injective' := λ f g h, by cases f; cases g; congr' }
 
+/-- Helper instance for when there's too many metavariables to apply `to_fun.to_coe_fn`. -/
 instance : has_coe_to_fun (seminorm 𝕜 E) (λ _, E → ℝ) := ⟨λ p, p.to_fun⟩
 
-@[ext] lemma ext {p q : seminorm 𝕜 E} (h : (p : E → ℝ) = q) : p = q :=
-begin
-  cases p,
-  cases q,
-  have : p_to_fun = q_to_fun := h,
-  simp_rw this,
-end
+@[ext] lemma ext {p q : seminorm 𝕜 E} (h : ∀ x, (p : E → ℝ) x = q x) : p = q := fun_like.ext p q h
+
+instance : has_zero (seminorm 𝕜 E) :=
+⟨{ to_fun    := 0,
+  smul'     := λ _ _, (mul_zero _).symm,
+  triangle' := λ _ _, eq.ge (zero_add _) }⟩
+
+@[simp] lemma coe_zero : ⇑(0 : seminorm 𝕜 E) = 0 := rfl
+
+instance : inhabited (seminorm 𝕜 E) := ⟨0⟩
 
 variables (p : seminorm 𝕜 E) (c : 𝕜) (x y : E) (r : ℝ)
 
 protected lemma smul : p (c • x) = ∥c∥ * p x := p.smul' _ _
 protected lemma triangle : p (x + y) ≤ p x + p y := p.triangle' _ _
 
+-- TODO: define `has_Sup` too, from the skeleton at
+-- https://github.com/leanprover-community/mathlib/pull/11329#issuecomment-1008915345
+noncomputable instance : has_sup (seminorm 𝕜 E) :=
+{ sup := λ p q,
+  { to_fun := p ⊔ q,
+    triangle' := λ x y, sup_le
+      ((p.triangle x y).trans $ add_le_add le_sup_left le_sup_left)
+      ((q.triangle x y).trans $ add_le_add le_sup_right le_sup_right),
+    smul' := λ x v, (congr_arg2 max (p.smul x v) (q.smul x v)).trans $
+      (mul_max_of_nonneg _ _ $ norm_nonneg x).symm } }
+
+@[simp] lemma coe_sup (p q : seminorm 𝕜 E) : ⇑(p ⊔ q) = p ⊔ q := rfl
+
+instance : partial_order (seminorm 𝕜 E) :=
+  partial_order.lift _ fun_like.coe_injective
+
+lemma le_def (p q : seminorm 𝕜 E) : p ≤ q ↔ (p : E → ℝ) ≤ q := iff.rfl
+lemma lt_def (p q : seminorm 𝕜 E) : p < q ↔ (p : E → ℝ) < q := iff.rfl
+
+noncomputable instance : semilattice_sup (seminorm 𝕜 E) :=
+function.injective.semilattice_sup _ fun_like.coe_injective coe_sup
```

Comment:
Here's the algebraic structure of seminorms. For the scalar one, you need to import `data.real.nnreal` and `open_locale nnreal`. It should work more generally but unfortunately there's something wrong within the order hierarchy.
```suggestion
function.injective.semilattice_sup _ fun_like.coe_injective coe_sup

instance : has_add (seminorm 𝕜 E) :=
⟨λ p q, { to_fun    := p + q,
          smul'     := λ c x, by rw [pi.add_apply, pi.add_apply, p.smul, q.smul, mul_add],
          triangle' := λ x y,
            (add_le_add (p.triangle _ _) $ q.triangle _ _).trans (add_add_add_comm _ _ _ _).le }⟩

@[simp] lemma coe_add (p q : seminorm 𝕜 E) : ⇑(p + q) = p + q := rfl

instance : add_comm_monoid (seminorm 𝕜 E) := fun_like.coe_injective.add_comm_monoid _ rfl coe_add

instance : ordered_cancel_add_comm_monoid (seminorm 𝕜 E) :=
fun_like.coe_injective.ordered_cancel_add_comm_monoid _ rfl coe_add

-- TODO: Generalize to `[canonically_ordered_comm_semiring 𝕝] [module 𝕝 ℝ] [ordered_smul 𝕝 ℝ]`
--  `[smul_comm_class 𝕝 ℝ ℝ]`
-- This currently doesn't work because `canonically_ordered_comm_semiring 𝕝` doesn't imply
-- `ordered_semiring 𝕝`
instance : has_scalar ℝ≥0 (seminorm 𝕜 E) :=
⟨λ a p, { to_fun    := a • p,
          smul'     := λ c x, by rw [pi.smul_apply, pi.smul_apply, mul_smul_comm, p.smul],
          triangle' := λ x y, (smul_le_smul_of_nonneg (p.triangle x y) $ nnreal.coe_nonneg _).trans
            (smul_add _ _ _).le }⟩

@[simp] lemma coe_smul (a : ℝ≥0) (p : seminorm 𝕜 E) : ⇑(a • p) = a • p := rfl
```
I can also PR it myself if you feel like this is too much new material.

Reply:
Oh whoops. I guess you can golf it then.